src/main/java/org/opensearch/neuralsearch/processor/TextEmbeddingProcessor.java

```diff
@@ -0,0 +1,278 @@
+            for (Map.Entry<String, Object> entry : ((Map<String, Object>) value).entrySet()) {
+                processMapTypeVectorOutput(
+                    entry.getKey(),
+                    entry.getValue(),
+                    modelTensorList,
+                    indexWrapper,
+                    (Map<String, Object>) sourceAndMetadataMap.get(String.valueOf(targetKey))
+                );
+            }
+        } else if (value instanceof String) {
+            sourceAndMetadataMap.put(targetKey, modelTensorList.get(indexWrapper.index++));
+        } else if (value instanceof List) {
+            sourceAndMetadataMap.put(targetKey, processListOut((List<String>) value, modelTensorList, indexWrapper));
+        }
+    }
+
+    private List<Map<String, List<Float>>> processListOut(
+        List<String> value,
+        List<List<Float>> modelTensorList,
+        IndexWrapper indexWrapper
+    ) {
+        List<Map<String, List<Float>>> numbers = new ArrayList<>();
+        for (String strValue : value) {
+            if (StringUtils.isNotBlank(strValue)) {
+                numbers.add(ImmutableMap.of(LIST_TYPE_NESTED_MAP_KEY, modelTensorList.get(indexWrapper.index++)));
+            }
+        }
+        return numbers;
+    }
+
+    private static void validateEmbeddingFieldsType(IngestDocument ingestDocument, Map<String, Object> embeddingFields) {
+        Map<String, Object> sourceAndMetadataMap = ingestDocument.getSourceAndMetadata();
+        for (Map.Entry<String, Object> entry : embeddingFields.entrySet()) {
+            Object obj = sourceAndMetadataMap.get(entry.getKey());
+            if (obj != null) {
+                String key = entry.getKey();
+                Class<?> clz = obj.getClass();
+                if (List.class.isAssignableFrom(clz) || Map.class.isAssignableFrom(clz)) {
+                    checkListElementsType(obj, key);
+                } else if (!String.class.isAssignableFrom(clz)) {
+                    throw new IllegalArgumentException("field [" + key + "] is neither string nor nested type, can not process it");
+                }
+            }
+        }
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    private static void checkListElementsType(Object obj, String key) {
+        if (List.class.isAssignableFrom(obj.getClass())) {
+            ((List) obj).stream().filter(Objects::nonNull).forEach(x -> checkListElementsType(x, key));
+        } else if (Map.class.isAssignableFrom(obj.getClass())) {
+            ((Map) obj).values().stream().filter(Objects::nonNull).forEach(x -> checkListElementsType(x, key));
+        } else if (!String.class.isAssignableFrom(obj.getClass())) {
+            throw new IllegalArgumentException("nested type field [" + key + "] has non-string type, can not process it");
+        }
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    public static final class Factory implements Processor.Factory {
```

Comment:
It seems this is a standard style of processor factory, this is an example: https://github.com/opensearch-project/OpenSearch/blob/d15795a7aca488c1fadb04b3c8d9f1a3b02e4056/modules/ingest-common/src/main/java/org/opensearch/ingest/common/SetProcessor.java#L115

Reply:
This is the opensearch style, because the factory code is little and the logic of processor creation is extremely related to the ingestion logic.

Reply:
Just a quick comment its not an OpenSearch style. Given the size of Processor class which you have provided as example it might be fine to put in the same file, but the current TextEmbeddingProcessor class is big enough which already makes it difficult to read. So, as best practice let's move Factory class out of this file. 

You can put the class under `src/main/java/org/opensearch/neuralsearch/processor/factory`.